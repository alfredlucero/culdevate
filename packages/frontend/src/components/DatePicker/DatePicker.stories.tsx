import React, { useState } from "react";
import { storiesOf } from "@storybook/react";
import DatePicker from "./index";

const StatefulDefaultDatePicker: React.FC = () => {
  const [dateValue, setDateValue] = useState<Date | null>(new Date());

  const handleDateChange = (date: Date | null) => {
    setDateValue(date);
  };

  return <DatePicker id="defaultDatePicker" selected={dateValue} onChange={handleDateChange} autoComplete="off" />;
};

storiesOf("Common/Date Picker", module)
  .add("Default Stateful Example", () => <StatefulDefaultDatePicker />)
  .add("Label + Text Info + Placeholder", () => (
    <DatePicker
      id="labelTextPlaceholder"
      selected={null}
      onChange={() => {}}
      label="Start Date"
      textInfo="When did you start this job?"
      placeholder="Job Start Date"
    />
  ))
  .add("Required + Error Info + Selected", () => (
    <DatePicker
      id="requiredErrorSelected"
      selected={null}
      onChange={() => {}}
      label="Start Date"
      errorInfo="Start Date is required"
      valid={false}
    />
  ))
  .add("Disabled + Selected", () => (
    <DatePicker id="disabledSelected" selected={new Date()} onChange={() => {}} label="Start Date" disabled={true} />
  ))
  .add("Month Year Picker", () => (
    <DatePicker
      id="monthYear"
      selected={new Date()}
      onChange={() => {}}
      showMonthYearPicker={true}
      dateFormat="MM/yyyy"
    />
  ))
  .add("Quarter Picker", () => (
    <DatePicker
      id="quarter"
      selected={new Date()}
      onChange={() => {}}
      showQuarterYearPicker={true}
      dateFormat="yyyy QQQ"
    />
  ));
