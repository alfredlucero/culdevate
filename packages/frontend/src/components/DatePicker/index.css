/* TODO: override calendar styles here */
