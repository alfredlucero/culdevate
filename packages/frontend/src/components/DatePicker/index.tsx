import React from "react";
import ReactDatePicker, { ReactDatePickerProps } from "react-datepicker";
import cn from "classnames";
import { CommonProps } from "../commonProps";
import "react-datepicker/dist/react-datepicker.css";
import "./index.css";

export interface DatePickerProps extends ReactDatePickerProps, CommonProps {
  id: string;
  valid?: boolean;
  errorInfo?: string;
  textInfo?: string;
  label?: string;
  placeholder?: string;
}

const DatePicker: React.FC<DatePickerProps> = ({
  id,
  selected,
  onChange,
  disabled = false,
  required = false,
  valid = true,
  errorInfo = "",
  textInfo = "",
  label = "",
  placeholder = "",
  className = "",
  testId = "",
  ...restOfDatePickerProps
}) => {
  return (
    <div {...(className !== "" ? { className } : {})}>
      {label && (
        <label className="block text-gray-700 text-sm font-bold mb-2" htmlFor={id}>
          {label} {required ? "" : "(optional)"}
        </label>
      )}
      <ReactDatePicker
        id={id}
        selected={selected}
        onChange={onChange}
        disabled={disabled}
        required={required}
        customInput={
          <input
            type="text"
            className={cn(
              [
                "shadow",
                "appearance-none",
                "border",
                "w-full",
                "py-2",
                "px-3",
                "text-gray-700",
                "leading-tight",
                "focus:outline-none",
                "focus:shadow-outline",
                "rounded",
              ],
              !valid ? ["border-red-500"] : [],
              disabled ? ["bg-gray-200", "opacity-75"] : [],
            )}
            {...(testId !== "" ? { "data-testid": testId } : {})}
          />
        }
        placeholderText={placeholder}
        {...restOfDatePickerProps}
      />
      {valid && textInfo !== "" && <p className="text-gray-700 text-xs italic mt-2">{textInfo}</p>}

      {!valid && errorInfo !== "" && <p className="text-red-500 text-xs italic mt-2">{errorInfo}</p>}
    </div>
  );
};

export default DatePicker;
