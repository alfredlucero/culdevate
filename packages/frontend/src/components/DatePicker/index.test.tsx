import React from "react";
import { render } from "@testing-library/react";
import DatePicker from "./index";

describe("<DatePicker />", () => {
  const datePickerTestId = "datePickerTestId";

  test("should render without error", () => {
    const { container } = render(
      <DatePicker
        id="datePicker"
        selected={new Date("2020/02/03")}
        onChange={() => {}}
        className="extra-date-class"
        testId={datePickerTestId}
      />,
    );

    expect(container).toMatchSnapshot();
  });

  test("should render label if present", () => {
    const label = "Date Label";
    const { getByLabelText, getByTestId } = render(
      <DatePicker
        id="datePicker"
        selected={new Date("2020/02/03")}
        onChange={() => {}}
        testId={datePickerTestId}
        label={label}
        required={true}
      />,
    );

    expect(getByLabelText(label)).toBeVisible();
    expect(getByTestId(datePickerTestId)).toBeRequired();
  });

  test("should render label with optional if present and not required", () => {
    const label = "Date Label";
    const { getByLabelText, getByTestId } = render(
      <DatePicker
        id="datePicker"
        selected={new Date("2020/02/03")}
        onChange={() => {}}
        testId={datePickerTestId}
        label={label}
        required={false}
      />,
    );

    expect(getByLabelText(`${label} (optional)`)).toBeVisible();
    expect(getByTestId(datePickerTestId)).not.toBeRequired();
  });

  test("should render inline text info if valid and info present", () => {
    const textInfo = "Some text info";
    const { container } = render(
      <DatePicker
        id="datePicker"
        selected={new Date("2020/02/03")}
        onChange={() => {}}
        testId={datePickerTestId}
        textInfo={textInfo}
        errorInfo="Error info"
        valid={true}
      />,
    );

    expect(container).toHaveTextContent(textInfo);
  });

  test("should render inline error info if invalid and error present", () => {
    const errorInfo = "Error info";
    const { container } = render(
      <DatePicker
        id="datePicker"
        selected={new Date("2020/02/03")}
        onChange={() => {}}
        testId={datePickerTestId}
        textInfo="Text info"
        errorInfo={errorInfo}
        valid={false}
      />,
    );

    expect(container).toHaveTextContent(errorInfo);
  });

  test("should render placeholder given nothing selected", () => {
    const placeholder = "Placeholder";
    const { getByPlaceholderText } = render(
      <DatePicker
        id="datePicker"
        selected={null}
        onChange={() => {}}
        testId={datePickerTestId}
        placeholder={placeholder}
      />,
    );

    expect(getByPlaceholderText(placeholder)).toBeVisible();
  });

  test("should render disabled state if disabled", () => {
    const { getByTestId } = render(
      <DatePicker id="datePicker" selected={null} onChange={() => {}} testId={datePickerTestId} disabled={true} />,
    );

    expect(getByTestId(datePickerTestId)).toBeDisabled();
  });
});
