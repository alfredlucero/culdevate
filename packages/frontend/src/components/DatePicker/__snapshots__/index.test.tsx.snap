// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DatePicker /> should render without error 1`] = `
<div>
  <div
    class="extra-date-class"
  >
    
    <div
      class="react-datepicker-wrapper"
    >
      <div
        class="react-datepicker__input-container"
      >
        <input
          class="shadow appearance-none border w-full py-2 px-3 text-gray-700 leading-tight focus:outline-none focus:shadow-outline rounded"
          data-testid="datePickerTestId"
          id="datePicker"
          placeholder=""
          type="text"
          value="02/03/2020"
        />
      </div>
    </div>
  </div>
</div>
`;
